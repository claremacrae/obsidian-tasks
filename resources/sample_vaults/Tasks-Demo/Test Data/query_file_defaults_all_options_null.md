---
tasks_query_explain:
tasks_query_short_mode:
tasks_query_show_tree:
tasks_query_show_tags:
tasks_query_show_id:
tasks_query_show_depends_on:
tasks_query_show_priority:
tasks_query_show_recurrence_rule:
tasks_query_show_on_completion:
tasks_query_show_created_date:
tasks_query_show_start_date:
tasks_query_show_scheduled_date:
tasks_query_show_due_date:
tasks_query_show_cancelled_date:
tasks_query_show_done_date:
tasks_query_show_urgency:
tasks_query_show_backlink:
tasks_query_show_edit_button:
tasks_query_show_postpone_button:
tasks_query_show_task_count:
tasks_query_extra_instructions:
---

# query_file_defaults_all_options_null

- [ ] #task Task in 'query_file_defaults_all_options_null'
