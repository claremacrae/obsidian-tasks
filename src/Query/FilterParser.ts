import { DescriptionField } from './Filter/DescriptionField';
import { DoneDateField } from './Filter/DoneDateField';
import { DueDateField } from './Filter/DueDateField';
import { ExcludeSubItemsField } from './Filter/ExcludeSubItemsField';
import { HeadingField } from './Filter/HeadingField';
import { PathField } from './Filter/PathField';
import { PriorityField } from './Filter/PriorityField';
import { ScheduledDateField } from './Filter/ScheduledDateField';
import { StartDateField } from './Filter/StartDateField';
import { HappensDateField } from './Filter/HappensDateField';
import { RecurringField } from './Filter/RecurringField';
import { StatusField } from './Filter/StatusField';
import { TagsField } from './Filter/TagsField';
import { BooleanField } from './Filter/BooleanField';

import type { FilterOrErrorMessage } from './Filter/Filter';

const fieldTypes = [
    BooleanField,
    StatusField,
    RecurringField,
    PriorityField,
    HappensDateField,
    StartDateField,
    ScheduledDateField,
    DueDateField,
    DoneDateField,
    PathField,
    DescriptionField,
    TagsField,
    HeadingField,
    ExcludeSubItemsField,
];

export function parseFilter(filterString: string): FilterOrErrorMessage | null {
    for (const fieldType of fieldTypes) {
        const field = new fieldType();
        if (field.canCreateFilterForLine(filterString))
            return field.createFilterOrErrorMessage(filterString);
    }
    return null;
}
