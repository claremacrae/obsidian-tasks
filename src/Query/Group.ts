import type { Moment } from 'moment';
import type { Grouping, GroupingProperty } from '../Query';
import type { Task } from '../Task';
import { TaskGroups } from './TaskGroups';

/**
 * A naming function, that takes a Task object and returns the corresponding group property name
 */
type Grouper = (task: Task) => string;

/**
 * Implementation of the 'group by' instruction.
 */
export class Group {
    private static readonly groupDateFormat = 'YYYY-MM-DD dddd';

    /**
     * Group a list of tasks, according to one or more task properties
     * @param grouping 0 or more Grouping values, one per 'group by' line
     * @param tasks The tasks that match the task block's Query
     */
    public static by(grouping: Grouping[], tasks: Task[]): TaskGroups {
        return new TaskGroups(grouping, tasks);
    }

    /**
     * Return the Grouper functions matching the 'group by' lines
     * @param grouping 0 or more Grouping values, one per 'group by' line
     */
    public static getGroupersForAllQueryGroupings(grouping: Grouping[]) {
        const groupers: Grouper[] = [];
        for (const { property } of grouping) {
            const comparator = Group.groupers[property];
            groupers.push(comparator);
        }
        return groupers;
    }

    /**
     * Return the group names for a single task
     * @param groupers The Grouper functions indicating the requested types of group
     * @param task
     */
    public static getGroupNamesForTask(groupers: Grouper[], task: Task) {
        const groupNames = [];
        for (const grouper of groupers) {
            const groupName = grouper(task);
            groupNames.push(groupName);
        }
        return groupNames;
    }

    /**
     * Return a single property name for a single task.
     * A convenience method for unit tests.
     * @param property
     * @param task
     */
    public static getGroupNameForTask(
        property: GroupingProperty,
        task: Task,
    ): string {
        const grouper = Group.groupers[property];
        return grouper(task);
    }

    private static groupers: Record<GroupingProperty, Grouper> = {
        backlink: Group.groupByBacklink,
        context: Group.groupByContext,
        done: Group.groupByDoneDate,
        due: Group.groupByDueDate,
        filename: Group.groupByFileName,
        folder: Group.groupByFolder,
        happens: Group.groupByHappensDate,
        heading: Group.groupByHeading,
        path: Group.groupByPath,
        priority: Group.groupByPriority,
        recurrence: Group.groupByRecurrence,
        recurring: Group.groupByRecurring,
        referenceDateField: Group.groupByHappensField,
        root: Group.groupByRoot,
        scheduled: Group.groupByScheduledDate,
        start: Group.groupByStartDate,
        status: Group.groupByStatus,
        urgency: Group.groupByUrgency,
    };

    private static groupByUrgency(task: Task): string {
        const value = task.urgency.toFixed(2);
        return `Urgency ${value}`;
    }

    private static groupByPriority(task: Task): string {
        return `Priority ${task.priority}`;
    }

    private static groupByRecurrence(task: Task): string {
        if (task.recurrence !== null) {
            return task.recurrence!.toText();
        } else {
            return 'None';
        }
    }

    private static groupByRecurring(task: Task): string {
        if (task.recurrence !== null) {
            return 'Recurring';
        } else {
            return 'Not Recurring';
        }
    }

    private static groupByStartDate(task: Task): string {
        return Group.groupByDate(task.startDate, 'start');
    }

    private static groupByScheduledDate(task: Task): string {
        return Group.groupByDate(task.scheduledDate, 'scheduled');
    }

    private static groupByDueDate(task: Task): string {
        return Group.groupByDate(task.dueDate, 'due');
    }

    private static groupByDoneDate(task: Task): string {
        return Group.groupByDate(task.doneDate, 'done');
    }

    private static groupByHappensDate(task: Task): string {
        const referenceDate = Group.getReferenceDate(task);
        const referenceName = Group.getReferenceDateField(task);
        if (referenceDate) {
            return Group.groupByDate(referenceDate, referenceName);
        }
        return 'No happens date';
    }

    private static groupByHappensField(task: Task): string {
        return Group.getReferenceDateField(task);
    }

    private static getReferenceDateField(task: Task) {
        let referenceName = 'None';
        if (task.dueDate != null) {
            referenceName = 'Due';
        } else if (task.scheduledDate != null) {
            referenceName = 'Scheduled';
        } else if (task.startDate != null) {
            referenceName = 'Start';
        }
        return referenceName;
    }

    private static getReferenceDate(task: Task) {
        let referenceDate: Moment | null = null;
        if (task.dueDate != null) {
            referenceDate = task.dueDate;
        } else if (task.scheduledDate != null) {
            referenceDate = task.scheduledDate;
        } else if (task.startDate != null) {
            referenceDate = task.startDate;
        }
        return referenceDate;
    }

    private static groupByDate(date: moment.Moment | null, field: string) {
        if (date === null) {
            return 'No ' + field + ' date';
        }
        return date.format(Group.groupDateFormat);
    }

    private static groupByPath(task: Task): string {
        // Does this need to be made stricter?
        // Is there a better way of getting the file name?
        return task.path.replace('.md', '');
    }

    private static groupByFolder(task: Task): string {
        const path = task.path;
        const fileNameWithExtension = task.filename + '.md';
        const folder = path.substring(
            0,
            path.lastIndexOf(fileNameWithExtension),
        );
        if (folder === '') {
            return '/';
        }
        return folder;
    }

    private static groupByFileName(task: Task): string {
        // Note current limitation: Tasks from different notes with the
        // same name will be grouped together, even though they are in
        // different files and their links will look different.
        const filename = task.filename;
        if (filename === null) {
            return 'Unknown Location';
        }
        return filename;
    }

    private static groupByRoot(task: Task): string {
        const path = task.path.replace('\\', '/');
        const separatorIndex = path.indexOf('/');
        if (separatorIndex == -1) {
            return '/';
        }
        return path.substring(0, separatorIndex + 1);
    }

    private static groupByBacklink(task: Task): string {
        const linkText = task.getLinkText({ isFilenameUnique: true });
        if (linkText === null) {
            return 'Unknown Location';
        }
        return linkText;
    }

    private static groupByStatus(task: Task): string {
        return task.status;
    }

    private static groupByHeading(task: Task): string {
        if (
            task.precedingHeader === null ||
            task.precedingHeader.length === 0
        ) {
            return '(No heading)';
        }
        return task.precedingHeader;
    }

    private static groupByContext(task: Task): string {
        const description = task.description;
        const contextRegexp = /.* #context\/([^ ]+)/;
        const contextMatch = description.match(contextRegexp);
        if (contextMatch !== null) {
            return contextMatch[1];
        }
        return 'No context';
    }
}
