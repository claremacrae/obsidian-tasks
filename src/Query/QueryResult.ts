import type { TaskGroups } from './TaskGroups';
import type { TaskGroup } from './TaskGroup';

export class QueryResult {
    public readonly taskGroups: TaskGroups;

    constructor(groups: TaskGroups) {
        this.taskGroups = groups;
    }

    public get totalTasksCount(): number {
        return this.taskGroups.totalTasksCount();
    }

    public get groups(): TaskGroup[] {
        return this.taskGroups.groups;
    }
}
