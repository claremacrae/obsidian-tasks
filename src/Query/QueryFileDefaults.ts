import type { OptionalTasksFile, TasksFile } from '../Scripting/TasksFile';
import { Query } from './Query';

/**
 * Construct query instructions from Obsidian properties in the query file
 */
export class QueryFileDefaults {
    public source(queryFile: OptionalTasksFile) {
        if (!queryFile) {
            return '';
        }
        const instructions = [
            // Instructions are listed in the order that items are displayed in Tasks search results
            this.instruction(queryFile, 'tasks_query_explain', 'explain', ''),
            this.instruction(queryFile, 'tasks_query_short_mode', 'short mode', 'full mode'),
            this.showAndHide(queryFile, 'tasks_query_show_tree', 'tree'),

            // Fields that appear before date values:
            this.showAndHide(queryFile, 'tasks_query_show_tags', 'tags'),
            this.showAndHide(queryFile, 'tasks_query_show_id', 'id'),
            this.showAndHide(queryFile, 'tasks_query_show_depends_on', 'depends on'),
            this.showAndHide(queryFile, 'tasks_query_show_priority', 'priority'),
            this.showAndHide(queryFile, 'tasks_query_show_recurrence_rule', 'recurrence rule'),
            this.showAndHide(queryFile, 'tasks_query_show_on_completion', 'on completion'),

            // Date fields:
            this.showAndHide(queryFile, 'tasks_query_show_created_date', 'created date'),
            this.showAndHide(queryFile, 'tasks_query_show_start_date', 'start date'),
            this.showAndHide(queryFile, 'tasks_query_show_scheduled_date', 'scheduled date'),
            this.showAndHide(queryFile, 'tasks_query_show_due_date', 'due date'),
            this.showAndHide(queryFile, 'tasks_query_show_cancelled_date', 'cancelled date'),
            this.showAndHide(queryFile, 'tasks_query_show_done_date', 'done date'),

            // Elements of query results:
            this.showAndHide(queryFile, 'tasks_query_show_urgency', 'urgency'),
            this.showAndHide(queryFile, 'tasks_query_show_backlink', 'backlink'),
            this.showAndHide(queryFile, 'tasks_query_show_edit_button', 'edit button'),
            this.showAndHide(queryFile, 'tasks_query_show_postpone_button', 'postpone button'),
            this.showAndHide(queryFile, 'tasks_query_show_task_count', 'task count'),

            // Extra instructions
            this.addValue(queryFile, 'tasks_query_extra_instructions'),
        ];
        return instructions.filter((i) => i !== '').join('\n');
    }

    private instruction(queryFile: TasksFile, prop: string, trueValue: string, falseValue: string) {
        return (queryFile.hasProperty(prop) && (queryFile.property(prop) ? trueValue : falseValue)) || '';
    }

    private showAndHide(queryFile: TasksFile, prop: string, field: string) {
        return this.instruction(queryFile, prop, 'show ' + field, 'hide ' + field);
    }

    private addValue(queryFile: TasksFile, prop: string) {
        return queryFile.hasProperty(prop) ? queryFile.property(prop) || '' : '';
    }

    public query(queryFile: OptionalTasksFile) {
        return new Query(this.source(queryFile), queryFile);
    }

    /**
     * Return text that creates MetaBind widgets for users to edit query file defaults.
     */
    public metaBindPluginWidgets() {
        // This is initially hard-coded, though I intend to machine-generate it eventually.
        // Its text is embedded in the test vault and in the user guide.
        return `
short mode: \`INPUT[toggle:tasks_query_short_mode]\`
tree: \`INPUT[toggle:tasks_query_show_tree]\`
tags: \`INPUT[toggle:tasks_query_show_tags]\`
id: \`INPUT[toggle:tasks_query_show_id]\` depends on: \`INPUT[toggle:tasks_query_show_depends_on]\`
priority: \`INPUT[toggle:tasks_query_show_priority]\`
recurrence rule: \`INPUT[toggle:tasks_query_show_recurrence_rule]\` on completion: \`INPUT[toggle:tasks_query_show_on_completion]\`
start date: \`INPUT[toggle:tasks_query_show_start_date]\` scheduled date: \`INPUT[toggle:tasks_query_show_scheduled_date]\` due date: \`INPUT[toggle:tasks_query_show_due_date]\`
created date: \`INPUT[toggle:tasks_query_show_created_date]\` cancelled date: \`INPUT[toggle:tasks_query_show_cancelled_date]\` done date: \`INPUT[toggle:tasks_query_show_done_date]\`
urgency: \`INPUT[toggle:tasks_query_show_urgency]\`
backlink: \`INPUT[toggle:tasks_query_show_backlink]\`
edit button: \`INPUT[toggle:tasks_query_show_edit_button]\` postpone button: \`INPUT[toggle:tasks_query_show_postpone_button]\`
task count: \`INPUT[toggle:tasks_query_show_task_count]\`
extra instructions: \`INPUT[textArea:tasks_query_extra_instructions]\`
explain: \`INPUT[toggle:tasks_query_explain]\`
`;
    }
}
