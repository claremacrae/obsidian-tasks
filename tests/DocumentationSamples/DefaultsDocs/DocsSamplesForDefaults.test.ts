import type { Pos } from 'obsidian';

import { getTasksFileFromMockData } from '../../TestingTools/MockDataHelpers';
import query_file_defaults_all_options_null from '../../Obsidian/__test_data__/query_file_defaults_all_options_null.json';
import query_file_defaults_all_options_true from '../../Obsidian/__test_data__/query_file_defaults_all_options_true.json';
import { verifyWithFileExtension } from '../../TestingTools/ApprovalTestHelpers';
import { verifyMarkdown, verifyMarkdownForDocs } from '../../TestingTools/VerifyMarkdown';
import { QueryFileDefaults } from '../../../src/Query/QueryFileDefaults';

function extractFrontmatter(data: any) {
    const queryFile = getTasksFileFromMockData(data);
    const pos: Pos | undefined = queryFile.cachedMetadata.frontmatterPosition;
    return data.fileContents.slice(pos?.start.offset ?? 0, pos?.end.offset ?? 0);
}

describe('DocsSamplesForDefaults', () => {
    it('supported-properties-empty', () => {
        verifyWithFileExtension(extractFrontmatter(query_file_defaults_all_options_null), '.yaml');
    });

    it('supported-properties-full', () => {
        verifyWithFileExtension(extractFrontmatter(query_file_defaults_all_options_true), '.yaml');
    });

    it('meta-bind-widgets-include', () => {
        verifyMarkdownForDocs(new QueryFileDefaults().metaBindPluginWidgets());
    });

    it('meta-bind-widgets-snippet', () => {
        verifyMarkdown(new QueryFileDefaults().metaBindPluginWidgets());
    });
});
