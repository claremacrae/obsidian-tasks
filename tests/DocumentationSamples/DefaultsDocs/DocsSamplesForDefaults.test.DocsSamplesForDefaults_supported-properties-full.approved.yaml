---
tasks_query_explain: true
tasks_query_short_mode: true
tasks_query_show_tree: true
tasks_query_show_tags: true
tasks_query_show_id: true
tasks_query_show_depends_on: true
tasks_query_show_priority: true
tasks_query_show_recurrence_rule: true
tasks_query_show_on_completion: true
tasks_query_show_created_date: true
tasks_query_show_start_date: true
tasks_query_show_scheduled_date: true
tasks_query_show_due_date: true
tasks_query_show_cancelled_date: true
tasks_query_show_done_date: true
tasks_query_show_urgency: true
tasks_query_show_backlink: true
tasks_query_show_edit_button: true
tasks_query_show_postpone_button: true
tasks_query_show_task_count: true
tasks_query_extra_instructions: |-
  # press shift-return to add new lines
  not done
---
